forge/src/executor/inspector/debugger.rs

```diff
@@ -0,0 +1,202 @@
+    pub ic_map: BTreeMap<Address, BTreeMap<usize, usize>>,
+}
+
+impl Debugger {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Builds the instruction counter map for the given bytecode.
+    // TODO: Some of the same logic is performed in REVM, but then later discarded. We should
+    // investigate if we can reuse it
+    pub fn build_ic_map(&mut self, address: &Address, code: &Bytes) {
+        let mut ic_map: BTreeMap<usize, usize> = BTreeMap::new();
+
+        let mut i = 0;
+        let mut cumulative_push_size = 0;
+        while i < code.len() {
+            let op = code[i];
+            ic_map.insert(i, i - cumulative_push_size);
+            match OpCode::is_push(op) {
+                Some(push_size) => {
+                    // Skip the push bytes
+                    i += push_size as usize;
+                    cumulative_push_size += push_size as usize;
+                }
+                None => (),
+            }
+            i += 1;
+        }
+
+        self.ic_map.insert(*address, ic_map);
+    }
+
+    /// Enters a new execution context.
+    pub fn enter(&mut self, depth: usize, address: Address, creation: bool) {
+        self.head =
+            self.arena.push_node(DebugNode { depth, address, creation, ..Default::default() });
+    }
+
+    /// Exits the current execution context, replacing it with the previous one.
+    pub fn exit(&mut self) {
+        if let Some(parent_id) = self.arena.arena[self.head].parent {
+            let DebugNode { depth, address, creation, .. } = self.arena.arena[parent_id];
+            self.head =
+                self.arena.push_node(DebugNode { depth, address, creation, ..Default::default() });
+        }
+    }
+
+    /// Records a debug step in the current execution context.
+    // TODO: Interpreter is only taken as a mutable borrow here because `Interpreter::gas` takes
+    // `&mut self` by mistake.
+    pub fn record_debug_step(&mut self, interpreter: &mut Interpreter) {
+        let pc = interpreter.program_counter();
+        let push_size = OpCode::is_push(interpreter.contract.code[pc]).map(|size| size as usize);
+        let push_bytes = push_size.as_ref().map(|push_size| {
+            let start = pc + 1;
+            let end = start + push_size;
+            interpreter.contract.code[start..end].to_vec()
+        });
+
+        self.arena.arena[self.head].steps.push(DebugStep {
+            pc,
+            stack: interpreter.stack().data().clone(),
+            memory: interpreter.memory.clone(),
+            instruction: Instruction::OpCode(interpreter.contract.code[pc]),
+            push_bytes,
+            ic: *self
+                .ic_map
+                .get(&interpreter.contract().address)
+                .expect("no instruction counter map")
+                .get(&pc)
+                .expect("unknown ic for pc"),
+            // TODO: The number reported here is off
+            total_gas_used: interpreter.gas().spend(),
```

Comment:
The numbers are off compared to the debugger on master - it doesn't update as frequently and the numbers are higher. Requires more investigation